modules/battery_manager/ajax/add_or_edit_entry.php

```diff
@@ -0,0 +1,207 @@
+
+    // Return JSON representation of duplicate entry if it exists, null otherwise
+    if ($entry) {
+        return json_encode($entry);
+    } else {
+        return null;
+    }
+}
+
+/**
+ * Handle insertions into the test battery
+ *
+ * @throws DatabaseException
+ *
+ * @return void
+ */
+function addEntry()
+{
+    $db   =& \Database::singleton();
+    $user =& User::singleton();
+    if (!$user->hasPermission('battery_manager_edit')) {
+        header("HTTP/1.1 403 Forbidden");
+        exit;
+    }
+
+    // Retrieve values entered by user
+    $form_data = getFormData();
+
+    // Make sure new entry is active
+    $form_data['Active'] = 'Y';
+
+    // Check for duplicates on the back-end
+    if (checkForDuplicate() !== null) {
+        header('HTTP/1.1 400 Bad Request');
+        exit("There exists a duplicate entry in the Test Battery.");
+    } else {
+        // Add entry to Test Battery
+        try {
+            $db->insert('test_battery', $form_data);
+        } catch (DatabaseException $e) {
+            showError("Could not add entry to the Test Battery. Please try again!");
+        }
+    }
+}
+
+/**
+ * Handle updates in the test battery
+ *
+ * @throws DatabaseException
+ *
+ * @return void
+ */
+function editEntry()
+{
+    $db   =& \Database::singleton();
+    $user =& User::singleton();
+    if (!$user->hasPermission('battery_manager_edit')) {
+        header("HTTP/1.1 403 Forbidden");
+        exit;
+    }
+
+    // Get ID of edited entry
+    $entryID = $_POST['id'];
+
+    // Retrieve values entered by user
+    $form_data = getFormData();
+
+    // Get active status of edited entry
+    $form_data['Active'] = $_POST['active'];
+
+    // Check for duplicates on the back-end
+    if (checkForDuplicate() !== null) {
+        header('HTTP/1.1 400 Bad Request');
+        exit("There exists a duplicate entry in the Test Battery.");
+    } else {
+        // Update entry in Test Battery
+        try {
+            $db->update('test_battery', $form_data, ['ID' => $entryID]);
+        } catch (DatabaseException $e) {
+            showError(
+                "Could not update entry "+$entryID+" in the Test Battery."
+                ." Please try again!"
+            );
+        }
+    }
+}
+
+/**
+ * Retrieve values entered by user
+ *
+ * @return array
+ */
+function getFormData()
+{
+    $form_data = array(
+                  'Test_name'    => $_POST['instrument'] ?? null,
+                  'AgeMinDays'   => $_POST['ageMinDays'] ?? null,
+                  'AgeMaxDays'   => $_POST['ageMaxDays'] ?? null,
+                  'Stage'        => $_POST['stage'] ?? null,
+                  'SubprojectID' => $_POST['subproject'] ?? null,
+                  'Visit_label'  => $_POST['visitLabel'] ?? null,
+                  'CenterID'     => $_POST['forSite'] ?? null,
+                  'firstVisit'   => $_POST['firstVisit'] ?? null,
+                  'instr_order'  => $_POST['instrumentOrder'] ?? null,
+                 );
+
```

Comment:
Data need to be sanitized to prevent SQL injection. Had been flagged before but still not resolved.

Reply:
It would be much more robust to prevent SQL injection via prepared statements